### PR TITLE
ADR: URL Path Design

### PR DESCRIPTION
**Pull request by yordis**

### CODE REVIEW

**Comment by yordis** [comment]
https://tools.ietf.org/html/rfc3986#section-6.2.2.1

**Review by tleef — APPROVED** [review]

**Review by mboperator — APPROVED** [review]
Clearly written and learned something new. Will be following these guidelines moving forward. Thanks, +1

File: adrs/8702033095/README.md [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+Creating "perfect software" is close to impossible. The following suggestions
+may feel like suggestions because we don't have the full scope of the situation.
+
+Therefore, you **MAY** follow the following resolutions unless you have a
+technical limitation. You **MUST** assumed such rules de facto in conversations.
+
+You **MUST** use the following characters anywhere in the path:
+
+* U+0061 to U+007A, "a-z"
+* U+0030 to U+0039, "0-9"
```
Comment (tleef) [inline adrs/8702033095/README.md:39]: What about `.` (period) which are used for things like file names "data.csv" or IDs sometimes and `~` which I like to use to separate out admin paths sometimes e.g. `mysite.com/~/settings`

**Comment by manuphatak** [comment]
Looks great!

Question! Are these in scope:

- URL params 
  - Is it safe to say these rules apply to URL params and "Hash params"?
  - Is it worth calling out explicitly?
  - If they are in scope, are their opinions for nested objects that might use `[]`s and `.`s?
- Trailing slashes
  - There are a variety of toolkits out there; some tools match trailing slashes implicitly (meaning they're optional), and some toolkits require an explicit mention of trailing slashes.
  - Are trailing slashes in scope?

File: adrs/8702033095/README.md (hunk 1 above)
Comment (tleef) [reply adrs/8702033095/README.md:39]:
```
Characters that are allowed in a URI, but do not have a reserved purpose, are called unreserved. These include uppercase and lowercase letters, decimal digits, hyphen, period, underscore, and tilde.

  ALPHA  DIGIT  "-" / "." / "_" / "~"
```

**Comment by manuphatak** [comment]
Also, examples!  I'd love to see examples of how to use the `-` vs `_`

File: adrs/8702033095/README.md (hunk 1 above)
Comment (yordis) [reply adrs/8702033095/README.md:39]: @tleef shit good catch ... forgot if `/hello.json` was part of the URL Path or not.

**Review by manuphatak — APPROVED** [review]
Approved!

Although, I have to mention that I LOVE that Figma uses emojis in their URLs! :kek:

**Comment by yordis** [comment]
> Although, I have to mention that I LOVE that Figma uses emojis in their URLs! :kek:

Yeah ... lets pretend that we don't need that unless we `HAVE` to, put that under the 10% range

**Comment by yordis** [comment]
```js
encodeURIComponent("👩‍❤️‍💋‍👨")
'%F0%9F%91%A9%E2%80%8D%E2%9D%A4%EF%B8%8F%E2%80%8D%F0%9F%92%8B%E2%80%8D%F0%9F%91%A8'
```

@manuphatak I don't think it is a good idea to be passing emojis in the URL, the URL has a size limit, and look how long a simple emoji encoding could be, therefore, avoid it.

**Comment by yordis** [comment]
> URL params
> - Is it safe to say these rules apply to URL params and "Hash params"?
> - Is it worth calling out explicitly?

Yeah, this is only for the URL Path, query parameters are outside of the scope.

> - Is it worth calling out explicitly?

I think we shouldn't explain too many concepts in ADRs, obviously a balance, but it can't be an educational content to teach people about edge cases like those, but, I am not 100% sure either thou, take it with a grain of salt.

What do you propose?

**Comment by yordis** [comment]
> Trailing slashes
> There are a variety of toolkits out there; some tools match trailing slashes implicitly (meaning they're optional), and some toolkits require an explicit mention of trailing slashes.
Are trailing slashes in scope?

@manuphatak where is your head is at? I am not following, but it sounds like something outside of the scope.

Being said, we can collaborate over https://github.com/straw-hat-team/specifications eventually since I will be adding some rules around the OpenAPI file, so maybe we combine it with some ADR as well.

**Comment by manuphatak** [comment]
> > URL params
> > 
> > * Is it safe to say these rules apply to URL params and "Hash params"?
> > * Is it worth calling out explicitly?
> 
> Yeah, this is only for the URL Path, query parameters are outside of the scope.



That's fair!  On rereading, the ADR does say "path" in several places.  However, since we're talking about URLs, I think many people will be thinking of the entire thing and may or may not pick up on the "path" mention.

I would recommend a note in the ADR to clarify the scope.  Something like "These rules apply to URL Paths.  Query params have their own rules." 




> 
> > * Is it worth calling out explicitly?
> 
> I think we shouldn't explain too many concepts in ADRs, obviously a balance, but it can't be an educational content to teach people about edge cases like those, but, I am not 100% sure either thou, take it with a grain of salt.
> 
> What do you propose?


Trailing slashes popped in my head because 1) the title references "separators" and 2) one of the goals is to make sense of the differences between tooling.    However, I think "trailing slashes" are out of scope, and they could be explored in their own ADR if necessary.

**Pull request merged**